source/module_elecstate/module_charge/charge_init.cpp

```diff
@@ -53,61 +53,37 @@ void Charge::init_rho(elecstate::efermi& eferm_iout,
             {
                 std::stringstream ssc;
                 ssc << PARAM.globalv.global_readin_dir << "SPIN" << is + 1 << "_CHG.cube";
-                double& ef_tmp = eferm_iout.get_ef(is);
-                if (ModuleIO::read_cube(
-#ifdef __MPI
-                    & (GlobalC::Pgrid),
-#endif
+                if (ModuleIO::read_cube(GlobalC::Pgrid,
                     (PARAM.inp.esolver_type == "sdft" ? GlobalV::RANK_IN_STOGROUP : GlobalV::MY_RANK),
-                    is,
                     GlobalV::ofs_running,
-                    PARAM.inp.nspin,
                     ssc.str(),
                     this->rho[is],
-                    this->rhopw->nx,
-                    this->rhopw->ny,
-                    this->rhopw->nz,
```

Comment:
my personal choice would be retaining the number of grid points for this function, but "the number of grid points" I mean are after collection among all processors.

Reply:
Yes, I've considered about this way. 

But from the caller's perspective: they always holds local-sized data, I feel it better to do the parallel reduction secretively under `Pgrid`, rather than expose the reduction interfaces to the caller before writing files.

Reply:
@maki49 no it is not the context or debate that viewing from whose perspective, it is **one function can only do one thing**. Scattering and gathering are not read_ and write_cube should do

Reply:
say there is a class can be "init" by a cube, what is correct is,
1. use read_cube to get information from the external cube file, say you get "rho"
2. use another function, say "init" to use rho to do value assignment. If you need rho to be distributed, inside function body of init, call a third function distribute_on_grid or something to distribute it
3. in the function body of distribute_on_grid, use #ifdef __MPI to enclose all mpi related lines, and if without MPI, then it is an empty function

rather than doing those three things in one function

Reply:
sounds the `read_grid` is a quite specific implementation, it is okay to have it in not basic function case. But for read_cube, the one who programs it should make it as general as possible, or it will become a trouble (you have seen in some cases before).